wolfcrypt/src/port/maxim/max3266x.c

```diff
@@ -0,0 +1,896 @@
+        if ((hash== NULL) || (data == NULL))
+            return BAD_FUNC_ARG;
+        if (hash->size < hash->used+size) {
+            if (hash->msg == NULL) {
+                p = XMALLOC(hash->used+size, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            }
+            else {
+                #ifdef WOLFSSL_NO_REALLOC
+                p = XMALLOC(hash->used + size, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                if (p != NULL) {
+                    XMEMCPY(p, hash->msg, hash->used);
+                    XFREE(hash->msg, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                }
+                #else
+                p = XREALLOC(hash->msg, hash->used+size, NULL,
+                                DYNAMIC_TYPE_TMP_BUFFER);
+                #endif
+
+            }
+            if (p == NULL){
+                return -1;
+            }
+            hash->msg = p;
+            hash->size = hash->used+size;
+        }
+        XMEMCPY(hash->msg+hash->used, data, size);
+        hash->used += size;
+        if(hash->msg == NULL){
+            return BAD_FUNC_ARG;
+        }
+    }
+    return 0;
+}
+
+int wc_MXC_TPU_SHA_GetHash(wc_MXC_Sha *hash, unsigned char* digest,
+                                MXC_TPU_HASH_TYPE algo)
+{
+    int status;
+    status = wc_MXC_TPU_SHA_GetDigest(hash, digest, algo);
+    /* True Case that msg is an empty string */
+    if(status == 1){
+        return 0;
+    }
+    /* False Case where msg needs to be processed */
+    else if (status == 0){
+        status = wolfSSL_CryptHwMutexLock();
+        if(wc_MXC_error(&status) != 0){
+
+            return status;
+        }
+        MXC_TPU_Init(MXC_SYS_PERIPH_CLOCK_TPU);
+        MXC_TPU_Hash_Config(algo);
+        status = MXC_TPU_Hash_SHA((const char *)hash->msg, algo, hash->size,
+                                         (char *)digest);
+        WOLFSSL_MSG("SHA HW Acceleration Used");
+        wolfSSL_CryptHwMutexUnLock();
+        if(wc_MXC_error(&status) != 0){
+            WOLFSSL_MSG("SHA HW Error Occured");
+            return status;
+        }
+    }
+    /* Error Occured */
+    return status;
+}
+
+int wc_MXC_TPU_SHA_Final(wc_MXC_Sha *hash, unsigned char* digest,
+                                    MXC_TPU_HASH_TYPE algo)
+{
+    int status;
+    status = wc_MXC_TPU_SHA_GetHash(hash, digest, algo);
+    if (status != 0){
+        return status;
+    }
+    XFREE(hash->msg, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    status = wc_MXC_TPU_SHA_Init(hash);
+    if (status != 0){
+        return status;
+    }
+    return status;
+}
+
+int wc_MXC_TPU_SHA_Copy(wc_MXC_Sha* src, wc_MXC_Sha* dst)
+{
+    if (src == NULL || dst == NULL) {
+        return BAD_FUNC_ARG;
+    }
+    dst->used = src->used;
+    dst->size = src->size;
+    XMEMCPY(dst->hash, src->hash, sizeof(dst->hash));
+    return XMEMCMP(dst->hash, src->hash, sizeof(dst->hash));
+}
+
+void wc_MXC_TPU_SHA_Free(wc_MXC_Sha* hash)
+{
+    XFREE(hash->msg, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    hash->msg = NULL;
```

Comment:
Looks like `wc_MXC_TPU_SHA_Init` does this (`hash->msg = NULL;`) and compiler would likely remove this line.